snowdrop-core/src/Snowdrop/Core/ERoComp/Helpers.hs

```diff
@@ -1,217 +1,155 @@
-    )
-    => ChangeSet id value
-    -> ERoComp e id value ctx a
-    -> ERoComp e id value ctx a
+  :: forall xs e ctx a .
+    (HasException e CSMappendException, HasLens ctx (ChgAccumCtx ctx), Default (ChgAccum ctx))
+  => HChangeSet xs
+  -> ERoComp e ctx xs a
+  -> ERoComp e ctx xs a
 withModifiedAccumCtx chgSet comp = do
     ctxAcc <- getCAOrDefault @ctx . gett <$> ask
     newAccOrErr <- modifyAccum ctxAcc $ CAMChange chgSet
     case newAccOrErr of
-        Left err   -> throwLocalError err
-        Right (acc', _undo) ->
-            local ( lensFor @ctx @(ChgAccumCtx ctx) .~ CAInitialized @ctx acc' ) comp
+        Left err            -> throwLocalError err
+        Right (acc', _undo) -> local ( lensFor @ctx @(ChgAccumCtx ctx) .~ CAInitialized @ctx acc' ) comp
 
 initAccumCtx
-    :: forall e id value ctx a .
-    (HasException e StatePException, HasLens ctx (ChgAccumCtx ctx))
-    => ChgAccum ctx
-    -> ERoComp e id value ctx a
-    -> ERoComp e id value ctx a
+  :: forall xs e ctx a .
+     (HasException e StatePException, HasLens ctx (ChgAccumCtx ctx))
+  => ChgAccum ctx
+  -> ERoComp e ctx xs a
+  -> ERoComp e ctx xs a
 initAccumCtx acc' comp = do
     gett @_ @(ChgAccumCtx ctx) <$> ask >>= \case
         CAInitialized _ -> throwLocalError ChgAccumCtxUnexpectedlyInitialized
-        CANotInitialized ->
-            local ( lensFor @ctx @(ChgAccumCtx ctx) .~ CAInitialized @ctx acc' ) comp
-
-valid :: (Monoid a, Monad m) => m a
-valid = pure mempty
-
-validateIff :: forall e e1 m a . (Monoid a, MonadError e m, HasReview e e1) => e1 -> Bool -> m a
-validateIff e1 = bool (throwLocalError e1) valid
-
-validateAll
-    :: (Foldable f, MonadError e m, Container (f a))
-    => (Element (f a) -> e)
-    -> (Element (f a) -> Bool)
-    -> f a
-    -> m ()
-validateAll ex p ls = maybe (pure ()) (throwError . ex) (find (not . p) ls)
+        CANotInitialized -> local ( lensFor @ctx @(ChgAccumCtx ctx) .~ CAInitialized @ctx acc' ) comp
 
 ------------------------
--- Compute undo
+-- Cast and hoist
 ------------------------
 
-data StateModificationException id
-    = UnexpectedKeyExists id
-    | UnexpectedKeyAbsent id
-    deriving (Show)
+type UpCastableERo xs supxs =
+    ( HIntersectable supxs xs
+    , HUpCastableMap xs supxs
+    , HUpCastableSet xs supxs
+    , HUpCastableChSet xs supxs
+    )
+
+upcastDbAccess
+    :: forall xs supxs chgAcc a . UpCastableERo xs supxs
+    => DbAccess chgAcc xs a
+    -> DbAccess chgAcc supxs a
+upcastDbAccess (DbQuery s cont) =
+    DbQuery (hupcast @_ @xs @supxs s) (\resp -> cont (hintersect @supxs @xs resp s))
+upcastDbAccess (DbIterator pr foldf) = DbIterator (pr . hdowncast) foldf
+upcastDbAccess (DbModifyAccum acc md cont) = case md of
+    CAMChange cs -> DbModifyAccum acc (CAMChange $ hupcast @_ @xs @supxs cs) (cont . fmap (fmap downcastUndo))
+    CAMRevert undo -> DbModifyAccum acc (CAMRevert $ upcastUndo @xs @supxs undo) (cont . fmap (fmap downcastUndo))
+
+upcastEffERoComp
```

Comment:
We should be cautious with using this transformation as additional `ReaderT` is always added. Later I would think of finding ways to avoid this.

Reply:
he he, yep. it's nice consideration.

Reply:
maybe we should even create a issue